libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,284 @@
+					for (auto& var: assignment->variableNames)
+						if (m_currentFunctionMemorySlots->count(var.name))
+						{
+							uint64_t slot = m_currentFunctionMemorySlots->at(var.name);
+							var.name = m_nameDispenser.newName(var.name);
+							mstores.emplace_back(getMemoryLocation(slot), var.name);
+							tempDecl.variables.emplace_back(TypedName{loc, var.name, {}});
+						}
+					std::vector<Statement> result;
+					result.emplace_back(std::move(tempDecl));
+					result.emplace_back(std::move(_statement));
+					appendMemoryStores(result, loc, mstores);
+					return {std::move(result)};
+				}
+				else if (
+					auto varDecl = std::get_if<VariableDeclaration>(&_statement);
+					varDecl && m_currentFunctionMemorySlots && util::contains_if(varDecl->variables, [&](auto const& var) {
+						return m_currentFunctionMemorySlots->count(var.name);
+					})
+				)
+				{
+					auto loc = locationOf(_statement);
+					std::vector<Statement> result;
+					std::vector<std::pair<YulString, YulString>> mstores;
+					for (auto& var: varDecl->variables)
+						if (m_currentFunctionMemorySlots->count(var.name))
+						{
+							uint64_t slot = m_currentFunctionMemorySlots->at(var.name);
+							var.name = m_nameDispenser.newName(var.name);
+							mstores.emplace_back(getMemoryLocation(slot), var.name);
+						}
+					result.emplace_back(std::move(_statement));
+					appendMemoryStores(result, loc, mstores);
+					return {std::move(result)};
+				}
+				else
+				{
+					ASTModifier::visit(_statement);
+					return {};
+				}
+			}
+		);
+	}
+	virtual void visit(Expression& _expression)
+	{
+		if (
+			auto identifier = std::get_if<Identifier>(&_expression);
+			identifier && m_currentFunctionMemorySlots && m_currentFunctionMemorySlots->count(identifier->name)
+		)
+		{
+			auto loc = identifier->location;
+			_expression = FunctionCall {
+				loc,
+				Identifier{loc, "mload"_yulstring}, {
+					Literal {
+						loc,
+						LiteralKind::Number,
+						getMemoryLocation(m_currentFunctionMemorySlots->at(identifier->name)),
+						{}
+					}
+				}
+			};
+		}
+		else
+			ASTModifier::visit(_expression);
+	}
+private:
+	std::map<YulString, std::map<YulString, uint64_t>> const& m_memorySlots;
+	u256 m_reservedMemory;
+	NameDispenser& m_nameDispenser;
+	std::map<YulString, uint64_t> const* m_currentFunctionMemorySlots = nullptr;
+};
+
+}
+
+void MemoryEscalator::run(OptimiserStepContext& _context, Object& _object, bool _optimizeStackAllocation)
+{
+	auto functionStackErrorInfo = CompilabilityChecker::run(_context.dialect, _object, _optimizeStackAllocation);
+	if (functionStackErrorInfo.empty())
+		return;
+
+	yulAssert(dynamic_cast<EVMDialect const*>(&_context.dialect), "MemoryEscalator can only be run on EVMDialect objects.");
+	// TODO: make this more robust with value tracking and stepping through the outermost block
+	FunctionCall* memoryInit = findMemoryInit(_object);
+	if (!memoryInit)
+		return;
+	u256 reservedMemory = 0;
```

Comment:
Is this the initial value of the free memory pointer?

Reply:
Yes - this *is* a bit hacky, I'm not sure how best to do this. In construction code we already have reserved memory for the immutables and I can't really put something *before* the immutables, so I have to fetch whatever literal we wrote to the free memory pointer to know where *after* the reserved space for the immutables starts to use that as the start offset of memory variables...
I'll add more comments or clean things up before taking it out of draft mode in any case.

Reply:
I think it's perfectly fine. All code assumes that the free memory pointer is an opaque value and only memory after its value can be freely used.

We should be careful, though, that the optimizer does not somehow replace the `mload(0x40)` later in the main routine. So maybe this _should_ rather be a builtin.

Reply:
We could even make it a builtin with a *required literal* argument for now.